Processing Datasets that Span Multiple Data Collection Runs (#88)
HumanCellAtlas/dcp-community: rfcs/text/0000-multi-data-collection-dataset-processing.md

As a data consumer, I need to know that a data set is done so that I don't download and use it until it is version-complete.

## Detailed Design
A new concept of *data group* is added to the DCP data model to address these issues. A data group is defined as a set of specific versions of metadata and data that is complete and consistent by specified criteria. A *data group* is not created until all its contents are complete and submitted to the DSS.  Events are generated when a *data group* is created, updated, or deleted.

A data group has a symbolic scope type that specifies what is represented by the group, as well a the set of criteria by which it is complete.  Most of the above use cases will require a *project submission* scope that indicates submission to a project is complete.

Data groups are implemented as a new bundle type that contains a JSON file listing the FQIDs (UUIDs with versions) of all bundles in the data group. The existing DSS subscription mechanism is used for notifications.

> A new schema *data_group* will be created that contains the fields:

Reviewer: >A new schema *data_group* will be created

Where would this definition live? Would this live in the metadata schema? What core module would this be under? Project? File? A new core entity?

Reviewer 2: Good point.  Data groups are not part of the scientific metadata, so it doesn't belong in the current hierarchy.    A similar issue was raised by @hannes-ucsc previously in a discussion about having bundle manifest metadata.

I will add this to the issues to be resolved during implementation.  It is possible that DSS collections could be a better way to implement data groups, however, the DSS documentation is no specific enough to make this decision at this time.

## Application to analysis pipelines

While this RFC proposes a general mechanism, it was developed in response to the needs of analysis to co-process data that are in multiple assay bundles.  This section uses co-processing of multiple sequencing run from a given library as an example of the use of data groups to identify data that should be processed together.  A key concept behind this proposal is that upstream submission should not be defining how downstream analysis groups inputs for processing.  The analysis pipelines need to be able to group input based on ways that may not be defined at submission time.  New analysis pipelines should be able to group inputs in arbitrary ways without requiring changes to the way data is packaged.

### Identifying data to co-process
Pipelines that require grouping data from multiple sequencing assays (co-processing) subscribe to *PROJECT_SUBMISSIONS* events instead of per-bundle events. Unlike with the assay bundle per pipeline run, this makes the pipeline framework responsible for collecting assay bundles into pipeline runs.  Pipelines that do not require co-processing continue to subscribe to assay bundle events and are unaffected.

> For full background and a detailed proposal on how these libraries are represented in the metadata see the RFC (Representing sequencing library preparations in the HCA DCP metadata standard)[https://github.com/HumanCellAtlas/dcp-community/blob/mfreeberg-rfc-bundle-definitions/rfcs/text/0000-bundle-definition.md].

Reviewer: Looks like the RFC link is meant to be:

```suggestion
For full background and a detailed proposal on how these libraries are represented in the metadata see the RFC (Representing sequencing library preparations in the HCA DCP metadata standard)[https://github.com/HumanCellAtlas/dcp-community/blob/68f0c42fc9bd55297be8c241663721271c90b694/rfcs/text/0010-rfc-library-preparation.md].
```

The creation and update of *PROJECT_SUBMISSIONS* *data groups*, both primary and secondary, is the responsibility of the Ingest system.

![Figure 2](/rfcs/images/0000-multi-data-collection-data-processing-images/flow-of-information.png)
Figure 2: Flow of information in the proposed data model

#### Updates to *data groups*
Updates to *PROJECT_SUBMISSIONS* must result in updating of the associated *data groups* to trigger reprocessing. When part or the entirety of a project submission is updated the associated data group must be identified and updated with the new bundle versions if bundles have been replaced and/or with additional new bundles if bundles have been added. This is the responsibility of Ingest.  Analysis must handle the update to the *data group* by initiating only the pipelines that have been affected by the update and ensure that the output bundles of this analysis are updates to existing bundles.

*Data group* updates may result from creating new bundles to add additional types of data to a project, or updating bundles to change metadata or append data to data files (e.g. FASTQ top-up).

### Unresolved Questions

> * *Data group* concept relates directly to the UX dissuasion on *experiment data sets*.  This RFC and that work could be unified and replace PROJECT_SUBMISSIONS

Reviewer: ```suggestion
* *Data group* concept relates directly to the UX discussion on *experiment data sets*.  This RFC and that work could be unified and replace PROJECT_SUBMISSIONS
```

### DCP PR:

***Leave this blank until the RFC is approved** then the **Author(s)** must create a link between the assigned RFC number and this pull request in the format:*

`[dcp-community/rfc#](https://github.com/HumanCellAtlas/dcp-community/pull/<PR#>)`

# RFC: Processing Datasets that Span Multiple Bundles

## Summary
This RFC proposes a solution to allow processing of datasets that span multiple bundles. It addresses the general problem that the current DCP data model contains no representation of data set grouping and version-completeness beyond that of individual data bundles.  This results in a restriction that all data processing within the DCP is a one-to-one linear set of steps deriving one bundle from another.

> However, there are use cases where a giving processing step can require input from multiple input bundles,

Reviewer 3: ~~giving~~?

resulting in a full DAG rather than a linear graph. The DAG sub-graph must be defined and all data available to initiate processing.  A notion of data completeness is required to support the implementation of DAG processing.

> This RFC defines a general mechanism for grouping bundles, called *data groups*, to allow bundles to be grouped in new ways.   By providing a general method for grouping bundles, this proposal provides a mechanism for addressing various tasks that cross bundle boundaries.  This impacts analysis pipelines, algorithmic complexity, data consistency, data set quality control and data releases.  The current bundle grouping under consideration is a submission to a project.  However, the concept generalizes in a manner that other groupings can be defined to accommodate future needs.

Reviewer 4: Is per submission a viable grouping? It only corresponds to [a set of changes that a submitter gave us](https://docs.google.com/document/d/1qlgIROPK4qESy9-Lwvva1ZWJ3S5EdZ0HWJiGGvt2JRM/edit#heading=h.ghd04i50lpgp). Those changes could have any meaning from being a whole dataset to being a correction of a misspelling in an author's name.

Reviewer 2: Interesting question.  I think we need to go through some of the downstream use cases and see what is easier for the consumers.  It is possible.

If it is just the files/bundles in an update, then analysis would query outside of the bounds of the data group to determine what needs to processed.   If the DG corresponds to something larger that bounds what needs to get processed it might make the query easier.

My feeling is that the DG will be more understandable if it corresponds to some scientific entity (experiment, project, project of projects)  and then updates are updates to the DG rather than new DGs

## Author(s)
- [Nick Barkas](mailto:barkasn@broadinstitute.org)
- [Mark Diekhans](mailto:markd@ucsc.edu)

## Shepherd
- [Nick Barkas](mailto:barkasn@broadinstitute.org)

## Motivation

> The current DCP model of processing has a scope of one data bundle, which contain a single assay or analysis.  There is no mechanism to know when a group of bundles that need to be processed as a group incomplete and available.  This restriction that data processing is linear and based on ingestion packaging causes multiple problems, as outlined in this section.

Reviewer 3: _incomplete_ `-->` _is complete_?

#### Multi-input analysis 
Analysis pipelines may require input from multiple different assays. To perform such analyses, they must be triggered only when all input data is available. The current mechanism of bundle notifications is not effective when data spans multiple bundles because there is no clear set of rules to identify the bundles that need to be co-processed.  Limiting the downstream process to the scope of a single bundle requires the ingest process to define the scope of the processing without any foreknowledge of future processes needs.

> An immediate need to for analysis pipelines to process 10X V2 scRNA-seq datasets that span multiple bundles exists. The need for co-processing multiple bundles is not however limited to this scenario and extends to any other NGS modality that involves repeated sequencing of the same library. Furthermore, the need to co-process data may extend to other data modalities that the project will accept in the future.

Reviewer 3: `need to for`

#### Data set quality control
An approach to quality control (QC) is to run a partial or full analysis on a subset of the data before doing a full analysis.  This requires defining an analyzable subset of the data to pass on to the processing pipelines. A QC data subset must meet the criteria that allow the QC analysis to run.

#### Data releases
A data release is a grouping of bundles meeting some criteria of consistency and quality that is considered of value to the users to be addressed as a whole.  Data releases require grouping of specific versions of data in ways that can not be defined before release time.  There may be data quality and project reasons for sub-setting data in a release a given manner.

> The DSS currently support functionality called *collections*, which some have suggested can be used to support releases.  However, documentation of the semantics and functional restrictions can not be found, so it is unclear if this mechanism is useful.

Reviewer 5: Just in case you missed it: documentation of collections is on the swagger page https://dss.dev.data.humancellatlas.org/

Reviewer 2: The swagger doc is more helpful than readthedocs but not really sufficient.  I have no idea what a metadata fragment is or why one would include it.  The limit on 1000 items per put/patch makes it very undesirable for something that can contain millions of items.

There is work to improve the DSS documentation, so hopefully, this will work itself out.

As a data consumer, I want to be able to find all data files that need to be processed together so that I can run my data processing pipeline. I do not want to be restricted in designing analysis based on how data was submitted.

As a data consumer, I want to be confident that the HCA DCP has correctly processed all data files that belong together so that I know the matrices I receive have been generated correctly.

As a data consumer, I need to know that a data set is done so that I don't download and use it until it is version-complete.

## Detailed Design
A new concept of *data group* is added to the DCP data model to address these issues.  A data group is defined as a set of specific versions of metadata and data that is complete and consistent by specified criteria.  A *data group* is not created until all its contents are complete.  Events are generated when a *data group* is created, updated, or deleted.

A data group has a symbolic scope type that specifies what is represented by the group, as well a the set of criteria by which it is complete.  Most of the above use cases will require a *project submission* scope that indicates submission to a project is complete.

> Data groups are implemented as a new bundle type that contains a JSON file listing the FQIDs (UUIDs with versions) of all bundles that part of the data groups and the existing DSS subscription mechanism for notifications.

Reviewer 3: I cannot parse this sentence.

Reviewer 6: As written (assuming you meant to add an "are" after "bundles"), this seems like a recursive definition.

A new schema *data_group* will be created that contains the fields:

> - *scope* - Symbolic name of the scope, for example, *PROJECT_SUBMISSION*.

Reviewer 3: Is there a difference between the concepts of `project` and `project submission`?

Reviewer 4: Yes, described in [this doc](https://docs.google.com/document/d/1qlgIROPK4qESy9-Lwvva1ZWJ3S5EdZ0HWJiGGvt2JRM/edit#).

- *bundle_fqids* - List of FQIDs of bundles that are in the group.

> Initially, only the *PROJECT_SUBMISSION* scope will be implemented. *PROJECT_SUBMISSION* will indicate that the entirety of single project submission is version-complete and will be generated by ingesting on after all data bundles are committed to the DSS.  This is indicated that data in the project submission should be processed by downstream components.  Ingest is responsible for creating *PROJECT_SUBMISSION* *data groups* after the relevant bundles have been created.  Figure 1 shows a diagram of a project with multiple *data groups*.

Reviewer 4: In a conversation with Nick, the original proposal was to create a PROJECT_SUBMISSION notification (using your language) every time a submission happened for that project. The notification would be a data group (nee DAPS bundle) sent through the datastore. However, a particular submission could just be a part of a dataset, or even just something like all the donors when the user will add other metadata later.. In our original conversation we decided (maybe implictly by some of us) that it didn't matter - you'd just get whatever the project happened to be - it's [just some collection of data and metadata](https://docs.google.com/document/d/1qlgIROPK4qESy9-Lwvva1ZWJ3S5EdZ0HWJiGGvt2JRM/edit#). But it sounds like you need project to be a richer concept. What would that be?

Really, my feeling is that to make this work as you're describing, ingests needs new first class entities to model data groups directly rather than trying to fit these onto submissions or projects. Those concepts would need to fit into the submitter's mental model which is where it becomes a UX question (@gabsie).

Reviewer 4: I see this is more or less your first unresolved question where data group may may to experiment data sets.

Reviewer 2: Yes, I believe it might be very useful to map this to a scientific concept and represent it in the metadata.  Experiment seems to be the right concept, however, that needs to be carefully defined.

Reviewer 4: Yes, I would say a new entity is critical for this work to proceed unless we come up with a suitable definition for 'project'.

Reviewer 2: The current definition of DGs would allow this all to work independently of the experimental,
however, I feel like it will be a lot less understandable that way.  Another layer of confusion.

![Figure 1](/rfcs/images/0000-multi-data-collection-data-processing-images/project-submission-in-project.png)

## Application to analysis pipelines

While this RFC proposes a general mechanism, it was developed in response to the needs of analysis to co-process data that are in multiple assay bundles.  This section uses co-processing of multiple sequencing run from a given library as an example of the use of data groups to identify data that should be processed together.  A key concept behind this proposal is that upstream submission should not be defining how downstream analysis groups input for processing.  The analysis pipelines need to be able to group input based on ways that may not be defined at submission time.  New analysis pipelines should be able to group inputs in arbitrary ways without requiring changes to the way data is packaged.

### Identifying data to co-process

> Pipelines that require grouping data from multiple sequencing assays (co-processing) subscribe to *PROJECT_SUBMISSIONS* events instead of per-bundle events. Unlike with the assay bundle per pipeline run, this makes the pipeline framework responsible for collecting assay bundles into pipeline runs.  Pipelines that do not require co-processing continue to subscribe to assay bundle events and are unaffected.

Reviewer 3: Given the general possibility of top-up sequencing, isn't _every_ pipeline a project pipeline?

Reviewer 2: good point @mckinsel 

@justincc, how is top-up sequencing handled?

Reviewer 7: Top up would require an update to the primary data bundle to add additional fastqs or it'd be what Marcus says, every pipeline would be a project pipeline and would get a notification that new data has arrived for that sample and has to be re-run with all the data in mind. This does present a complex graph where old analysis results are for one set of inputs and new analysis results are for an aggregate.

Reviewer 4: @kbergin So these are new FASTQ files that should be added to an existing process that produces cell_suspension -> FASTQs? In a library preparation world, those FASTQs would come from the existing library preparation? If so, this is indeed adding files to an existing bundle. How common is this case? From other conversations I thought adding files to existing bundles was low-event/low importance. If this is wrong then we may need to adjust implementation priorities.

Reviewer 7: I could be interpreting top-up sequencing in a different way than @mckinsel intended.

At the Broad, a top up sequencing is used to hit sequencing coverage on a sample before the data is delivered to the customer that paid for it. In our world, the customer only ever receives one aggregated BAM that has already been processed and contains all of the data from multiple sequencing events in one BAM. If they were to revert this to an unaligned raw FASTQ, it would be one submission event from HCA's perspective.

If a data submitter has data for a sample and submits it and then receives more data for the same sample, it could be a few different scenarios. In general, it implies more sequencing of the same library preparation. In my experience in the lab, sometimes a new library preparation from the same sample is created to perform the additional sequencing to top up. I'm not sure if that's applicable to droplet based single cell experiments cc @barkasn 

I imagine that for the forseeable future, as we are receiving data from the data submitters after they have likely already performed processing on it from their sequencing core or group, we won't receive as much top up sequencing after the fact. It's probably more likely that we'll have a submitter just forget to include a fastq in their initial data submission that was there from the beginning and was not a top up that arrived later.

Does that explanation help or confuse further?

Reviewer 2: Interesting, that is different than my experience with Broad for TCGA where the goal was to get the data out quickly.  They call this BAM top-off, an I *think* that was getting a replacement BAM with the existing read groups plus the new read groups.  This would allow downstream analysis tools to maintain their assumptions about all data being in one BAM

Reviewer 4: This is helpful @kbergin. Do you have any feeling for how likely it is that a submitter forgets a FASTQ. Feeling here was that we really wouldn't expect to see that very often.

Reviewer 8: > Pipelines that do not require co-processing continue to subscribe to assay bundle events and are unaffected.

@diekhans @barkasn is this still true? From our conversation earlier this week, my understanding was that you could only create a subscription that receives all project submission events, unlike the existing subscription queries that allow you to match on a particular kind of data. This means we would only have one subscription that notifies us of every complete submission, and depending on the contents we would have to determine what pipeline to run. So if we received a project submission event consisting of SmartSeq2 data, but we still had a SmartSeq2-specific subscription we would trigger the analysis twice. 

If there was a way to create a more specific subscription to project submission events (e.g. give me projects that contain everything except SmartSeq2), then we would be able to subscribe to both assay bundle events and project submission events.

Reviewer 2: While processing is still on a per-bundle basis, the input grouping needs to be passed as an output data group for the needs of the downstream consumers like Azul.

Thanks for point out the outdate text, it has been updated.

> For full background and a detailed proposal on how these libraries are represented in the metadata see the RFC (Representing sequencing library preparations in the HCA DCP metadata standard)[https://github.com/HumanCellAtlas/dcp-community/blob/mfreeberg-rfc-bundle-definitions/rfcs/text/0000-bundle-definition.md].

Reviewer 5: ```suggestion
For full background and a detailed proposal on how these libraries are represented in the metadata see the RFC [Representing sequencing library preparations in the HCA DCP metadata standard](https://github.com/HumanCellAtlas/dcp-community/blob/mfreeberg-rfc-bundle-definitions/rfcs/text/0000-bundle-definition.md).
```

An analysis pipeline event handler component will be created that is responsible for partitioning a *data group* into co-processing units, dispatching the pipelines, and tracking completion. 

![Figure 2](/rfcs/images/0000-multi-data-collection-data-processing-images/flow-of-information.png)

#### Updates to *data groups*
Updates to PROJECT_SUBMISSIONS must result in updating of the associated *data groups* to trigger reprocessing. When part or the entirety of a project submission is updated the associated data group must be identified and updated with the new bundle versions if bundles have been replaced and/or with additional new bundles if bundles have been added. This is the responsibility of Ingest. Analysis must handle the update to the *data group* by initiating only the pipelines that have been affected by the update and ensure that the output bundles of this analysis are updates to existing bundles.

> ### Unresolved Questions

Reviewer 4: I think it is very important to think about how this will stretch beyond PROJECT_SUBMISSION. It doesn't need to be high detail but I would need a warm feeling that we are not heading down a development path that can't cater to later requirements.

Reviewer 4: As below, on reconsideration I don't think is too important to answer now as I suspect PROJECT_SUBMISSION is needed in pretty much all circumstances.

> ### Unresolved Questions

Reviewer 4: Could you consider whether much of this can be implemented by subscriptions to the query service? This would still need to be supplemented by ingest signals of submission but it may allow more flexibility for future requirements, where those can be catered for by subscription requirements rather than new data model entities in ingest, with all the work and coordination that involves.

Reviewer 2: I suspect it could be implemented by query service events, although it is imaged that the QS will use project data groups to populate the database.

Without understanding more about how QS events will integrate into the system, using the DSS event mechanism seems like a conservative step.

@MDunitz @kislyuk, any thoughts?

Reviewer 4: I think this is still and important question but even with this, some submission finished event (here corresponding to PROJECT_SUBMISSION) would still be necessary. Hence, I don't think it needs to be answered here.

> * *Data group* concept relates directly to the planned UX on on *experiment data sets*.  This RFC and that work should be unified.

Reviewer 4: Hey @gabsie do you have a link to this work?

* How *data groups* are to used group results in secondary analysis has not yet been defined.
* Could a *data group* be a DSS collection?  Collections are poorly document, so it is unclear if they can fit the needs.
* Could a *data group* be a bundle that contains the *data group* files rather than a list of bundles?  Can the DSS handle bundles with a very large number of files?
* How does this framework interact with new modalities we are likely to encounter in the future. A good example of this is imaging data?

## Dependencies

> Implementation of this RFC is dependent on (RFC: Bundle Types)[https://github.com/HumanCellAtlas/dcp-community/pull/86].

Reviewer 5: ```suggestion
Implementation of this RFC is dependent on [RFC: Bundle Types](https://github.com/HumanCellAtlas/dcp-community/pull/86).
```

### DCP PR:

***Leave this blank until the RFC is approved** then the **Author(s)** must create a link between the assigned RFC number and this pull request in the format:*

`[dcp-community/rfc#](https://github.com/HumanCellAtlas/dcp-community/pull/<PR#>)`

# RFC: Data Groups: Processing Datasets that Span Multiple Bundles

## Summary
This RFC proposes a solution to allow processing of datasets that span multiple bundles. It addresses the general problem that the current DCP data model contains no representation of data set grouping and version-completeness beyond that of individual data bundles.  The current DCP model is restricted in that all data processing can be a one-to-one linear sequence of steps deriving one bundle from another.

> However, there are use cases where a given processing step can require input from multiple input bundles, resulting in a full directed acyclic graph (DAG) rather than a linear graph. The relevant DAG sub-graph must be defined and all data that constitute is must be available to correctly initiate processing.  A notion and implemented representation of data completeness is required to support DAG processing.

Reviewer 9: typo: "all data that constitute is"

> This RFC defines a general mechanism for grouping bundles, here called *data groups*. By providing a general method for grouping bundles, this proposal provides a mechanism for addressing various tasks that cross bundle boundaries.  This has impact on analysis pipelines, algorithmic complexity of data processing, data consistency, and data set quality control.  The current bundle grouping, described in detail here, is a submission to a project.  However, the concept generalizes in a manner that other groupings can be defined to accommodate future needs.

Reviewer 9: Is the phrase "submission to a project" the same as "subset of project bundles"? If so, the latter is easier to understand for readers not steeped in DCP jargon.

Reviewer 2: technically, these are not the same.  The produce submits data to the project, ingest create the bundles from the submitted data.

An immediate need for analysis pipelines to process 10X V2 scRNA-seq datasets that span multiple bundles exists. The need for co-processing multiple bundles is not however limited to this scenario and extends to any other NGS modality that involves repeated sequencing of the same library. Furthermore, the need to co-process data may extend to other data modalities that the project will accept in the future.

In the future, a need to run analysis processes where input sets span multiple projects may arise. A mechanism to specify processing data collections that are not restricted to a single project or submissions will be required to support this functionality.

#### Algorithmic complexity and performance

> Producing combined metadata or data from multiple assay bundles using the bundle event system results in O(N^2) algorithmic complexity, where N is the number of data bundles in some grouping. As each bundle event arrives, it must be combined with accumulated metadata from previous bundles. Additionally, there is a race condition on reading and writing the accumulated the results that must be handled as the order of event receipts is not guaranteed.

Reviewer 9: type: "the accumulated the results"

As a data consumer, I want to be confident that the HCA DCP has correctly processed all data files that belong together so that I know the matrices I receive have been generated correctly.

As a data consumer, I need to know that a data set is done so that I don't download and use it until it is version-complete.

## Detailed Design
A new concept of *data group* is added to the DCP data model to address these issues. A data group is defined as a set of specific versions of metadata and data that is complete and consistent by specified criteria. A *data group* is not created until all its contents are complete and submitted to the DSS.  Events are generated when a *data group* is created, updated, or deleted.

A data group has a symbolic scope type that specifies what is represented by the group, as well a the set of criteria by which it is complete.  Most of the above use cases will require a *project submission* scope that indicates submission to a project is complete.

> Data groups are implemented as a new bundle type that contains a JSON file listing the FQIDs (UUIDs with versions) of all bundles in the data group. The existing DSS subscription mechanism is used for notifications.  DSS collections were considered as a possible implementation, however they don't support subscriptions and are not indexed, so they are not a sufficient mechanism for implementing *data groups*.

Reviewer 9: While I agree with them, IMO the remarks about collections do not aid the clarity of this design.

Reviewer 2: good point, I moved the discussion of why collections were not used to Alternatives.

### Identifying data to co-process
Pipelines that require grouping data from multiple sequencing assays (co-processing) subscribe to *PROJECT_SUBMISSIONS* events instead of per-bundle events. Unlike with the assay bundle per pipeline run, this makes the pipeline framework responsible for collecting assay bundles into pipeline runs.  Pipelines that do not require co-processing continue to subscribe to assay bundle events and are unaffected.

For full background and a detailed proposal on how these libraries are represented in the metadata see the RFC (Representing sequencing library preparations in the HCA DCP metadata standard)[https://github.com/HumanCellAtlas/dcp-community/blob/68f0c42fc9bd55297be8c241663721271c90b694/rfcs/text/0010-rfc-library-preparation.md].

> The *data group* submission event does not define bundles to be co-processed, it indicates that data to be co-processed is contained within the *data group* and that that group fulfills version-completeness requirements. Co-processing will be initiated by pipelines based on a well-defined set of rules that is data modality-specific.  The partition of a submission *data group* into co-processing units may require multiple parameters. For instance, with 10X V2 3’ scRNA-seq is grouped by library, however, a submission my contain non-10X data modalities that need to be processed separately.

Reviewer 9: Types in last sentence.

### DCP PR:

***Leave this blank until the RFC is approved** then the **Author(s)** must create a link between the assigned RFC number and this pull request in the format:*

`[dcp-community/rfc#](https://github.com/HumanCellAtlas/dcp-community/pull/<PR#>)`

# RFC: Data Groups: Processing Datasets that Span Multiple Bundles

## Summary
This RFC proposes a solution to allow processing of datasets that span multiple bundles. It addresses the general problem that the current DCP data model contains no representation of data set grouping and version-completeness beyond that of individual data bundles.  The current DCP model is restricted in that all data processing can be a one-to-one linear sequence of steps deriving one bundle from another.

> However, there are use cases where a given processing step can require input from multiple input bundles, resulting in a full directed acyclic graph (DAG) rather than a linear graph. The relevant DAG sub-graph must be defined and all of its associated data available to correctly initiate processing.  A notion and implemented representation of data completeness is required to support DAG processing.

Reviewer 10: re: "full directed acyclic graph" What makes the DAG "full"? I am not familiar with the term mathematically.

Reviewer 2: over-enthusiastic use of adjectives. changed.

> This RFC defines a general mechanism for grouping bundles, here called *data groups*. By providing a general method for grouping bundles, this proposal provides a mechanism for addressing various tasks that cross bundle boundaries.  This has impact on analysis ,pipelines, algorithmic complexity of data processing, data consistency, and data set quality control.  The current bundle grouping, described in detail here, is a submission to a project.  However, the concept generalizes in a manner that other groupings can be defined to accommodate future needs.

Reviewer 10: Nit: comma between analysis and pipelines is misplaced.

> This RFC defines a general mechanism for grouping bundles, here called *data groups*. By providing a general method for grouping bundles, this proposal provides a mechanism for addressing various tasks that cross bundle boundaries.  This has impact on analysis ,pipelines, algorithmic complexity of data processing, data consistency, and data set quality control.  The current bundle grouping, described in detail here, is a submission to a project.  However, the concept generalizes in a manner that other groupings can be defined to accommodate future needs.

Reviewer 10: re: "future needs" What other future needs do we anticipate? Is this going to be common that we want to group bundles differently and what is the evidence?

Reviewer 2: No specific needs have been identified that are not already mentioned.  However, a goal of the design was to allow for other uses as long as there was no compelling reason not to allow for them.  

The evidence is that the initial design of the DCP made an oversimplified requirement that the only unit of processing and completeness was defined by the way the data was stored by the submission transaction.  This created the problems outlined in the RFC.

## Author(s)
- [Nick Barkas](mailto:barkasn@broadinstitute.org)
- [Mark Diekhans](mailto:markd@ucsc.edu)

## Shepherd
- [Nick Barkas](mailto:barkasn@broadinstitute.org)

## Motivation

> The current DCP model of processing has the scope of a single data bundle, which contains a single assay or analysis.  There is no mechanism to know when a group of bundles that need to be processed together is complete and available for further processing.  This results in the limitation that data processing is linear and is entirely based on ingestion packaging resulting in multiple problems, as outlined below.

Reviewer 10: On "completeness"... how does this align with the concept of completeness presented in the Project Completeness RFC (https://github.com/HumanCellAtlas/dcp-community/pull/115).

Reviewer 2: This is a low-level mechanism that is used to specify when a grouping of data is complete.  This could (hopefully would) be used to implement the tools and policy in RFC PR #115 .

If the refinement of #115 creates new requirements, this RFC will need to be amended.

Reviewer 10: Sounds good! Thanks for the explanation!

#### Multi-input analysis 

> Analysis pipelines may require input from multiple different assays. To perform such analyses, they must be triggered only when all input data are available. The current mechanism of bundle notifications is not effective when data spans multiple bundles because there is no clear set of rules to identify the bundles that need to be co-processed.  This limits the downstream process to the scope of a single bundle and requires Ingest to define the scope of all subsequent processing without foreknowledge of future needs.

Reviewer 10: Nit: "To perform such analyses, they must be triggered only when all input data *is* available."

Reviewer 2: fixed

> An immediate need for analysis pipelines to process 10X V2 scRNA-seq datasets that span multiple bundles exists. The need for co-processing multiple bundles is not however limited to this scenario and extends to any other NGS modality that involves repeated sequencing of the same library. Furthermore, the need to co-process data may extend to other data modalities that the project will accept in the future.

Reviewer 10: I'm not familiar with the term "NGS." Would you be able to elaborate further please?

Reviewer 2: NGS == next-generation sequencing.  I generalized the sentence to allow for non-sequencing technologies.

> An immediate need for analysis pipelines to process 10X V2 scRNA-seq datasets that span multiple bundles exists. The need for co-processing multiple bundles is not however limited to this scenario and extends to any other NGS modality that involves repeated sequencing of the same library. Furthermore, the need to co-process data may extend to other data modalities that the project will accept in the future.

Reviewer 10: Nit: Can we substitute "DCP" for the word "project" in the sentence "Furthermore, the need to co-process data may extend to other data modalities that the project will accept in the future." for clarity?

Reviewer 2: yes, that is much clearer.  Thanks.

> In the future, a need to run analysis processes where input sets span multiple projects may arise. A mechanism to specify processing data collections that are not restricted to a single project or submissions will be required to support this functionality.

Reviewer 10: Do we know this for sure?

Reviewer 2: No, we don't know for sure how experiments will be designed in the future.  Experience shows that data gets used in lots of new, creative ways.  Unless we have a good reason not to, such as making it much harder to implement, aiming for flexibility is wise.

Reviewer 10: SGTM!

#### Algorithmic complexity and performance

> Producing combined metadata or data from multiple assay bundles using the bundle event system results in O(N^2) algorithmic complexity, where N is the number of data bundles in some grouping. As each bundle event arrives, it must be combined with accumulated metadata from previous bundles. Additionally, there is a race condition on reading and writing the accumulated results that must be handled as the order of event receipts is not guaranteed.

Reviewer 10: Could you elaborate a bit more on what the race condition is here?

Reviewer 2: Building accumulated data file in S3 from contents in multiple bundles.  For instance, creating a project metadata TSV:

- bundle event A handler reads project TSV
- bundle event B handler reads project TSV
- bundle event A handler adds A metadata to project metadata
- bundle event A handler writes project TSV
- bundle event B handler adds B metadata to project metadata
- bundle event B handler writes project TSV, overwriting A's updates

This is also O(N^2)

One approach to addressing this is to merely queue all data bundle events. This, however, requires knowledge of completion so that process can be triggered, which is not addressed by the use of a queue. The algorithmic complexity of combining incoming notifications, impacts the data browser when creating project metadata TSVs and normalized JSON files, as well as the matrix service when generating pre-built project matrices.

#### Data consistency

> Consistency and completeness across a data set is an important attribute of an atlas. An incomplete data set may result in misinterpretation of data or missed discoveries. While a data set can be updated in the future, there is a need to know when the current version of a data set is fully available. This is referred to as *version-completeness*. Currently, the DCP has no way to indicate version-completeness beyond the bundle level. Neither the data browser nor any consumer API and have the information to indicate that a project submission is complete. This information is only available within Ingest, with no way to communicate submission status to other components.

Reviewer 10: This paragraph focuses mostly on completeness. What does consistency mean to you here?

Reviewer 2: Consistency means that all the bundles in the group are in the same update "state".  For example, in duplicated metadata was being updated.
  
Text modified to state this.  Thanks.

## Application to analysis pipelines

While this RFC proposes a general mechanism, it was developed in response to the needs of analysis to co-process data that are in multiple assay bundles.  This section uses co-processing of multiple sequencing run from a given library as an example of the use of data groups to identify data that should be processed together.  A key concept behind this proposal is that upstream submission should not be defining how downstream analysis groups inputs for processing.  The analysis pipelines need to be able to group input based on ways that may not be defined at submission time.  New analysis pipelines should be able to group inputs in arbitrary ways without requiring changes to the way data is packaged.

### Identifying data to co-process
Pipelines that require grouping data from multiple sequencing assays (co-processing) subscribe to *PROJECT_SUBMISSIONS* events instead of per-bundle events. Unlike with the assay bundle per pipeline run, this makes the pipeline framework responsible for collecting assay bundles into pipeline runs.  Pipelines that do not require co-processing continue to subscribe to assay bundle events and are unaffected.

> For full background and a detailed proposal on how these libraries are represented in the metadata see the RFC (Representing sequencing library preparations in the HCA DCP metadata standard)[https://github.com/HumanCellAtlas/dcp-community/blob/68f0c42fc9bd55297be8c241663721271c90b694/rfcs/text/0010-rfc-library-preparation.md].

Reviewer 10: Nit: formatting error.

Reviewer 2: fixed

A new schema *data_group* will be created that contains the fields:
- *scope* - Symbolic name of the scope, for example, *PROJECT_SUBMISSION*.
- *bundle_fqids* - List of FQIDs of bundles that are in the group.

See also:
- [The Ingest Submission Data Model](https://docs.google.com/document/d/1qlgIROPK4qESy9-Lwvva1ZWJ3S5EdZ0HWJiGGvt2JRM/edit#heading=h.8r5egyz5ly82) for a description of submissions.
- [Bundling in ingest](https://docs.google.com/document/d/1cGfolHMGEe1IKkXSBS75SAdSri9Np_oW6Z4S9-Ax1QY/edit#heading=h.xv0ag3c16x15) for details on how bundles are created.

Initially, only the *PROJECT_SUBMISSION* scope will be implemented. *PROJECT_SUBMISSION* will indicate that a single submission of data to a project is version-complete and will be generated by Ingest only after all data bundles are committed to the DSS. The generation of the *PROJECT_SUBMISSION* data group indicates that data in the submission to the project should be processed by downstream components.  Ingest is responsible for creating *PROJECT_SUBMISSION* *data groups* after the relevant bundles have been created.  Figure 1 shows a diagram of a project with multiple *data groups*.

> ![Figure 1](../images/0000-multi-data-collection-data-processing-images/project-submission-in-project.png)

Reviewer 4: This diagram confuses me. Wouldn't a PROJECT_SUBMISSION data group always contain bundles from the entire project?

Reviewer 11: Yeah, that confused me too

Reviewer 2: Multiple data groups are possible in a project.  It was clarified in the text that it makes sense to have primary  and analysis data groups  

However, the drawing is confusing and needs to be updated.

Reviewer 2: Text has been updated to make this clearer.

For full background and a detailed proposal on how these libraries are represented in the metadata see the RFC [Representing sequencing library preparations in the HCA DCP metadata standard](https://github.com/HumanCellAtlas/dcp-community/blob/68f0c42fc9bd55297be8c241663721271c90b694/rfcs/text/0010-rfc-library-preparation.md).

The *data group* submission event does not define bundles to be co-processed, it indicates that data to be co-processed is contained within the *data group* and that that group fulfills version-completeness requirements. Co-processing will be initiated by pipelines based on a well-defined set of rules that is data modality-specific.  The partition of a submission *data group* into co-processing units may require multiple parameters. For instance, with 10X V2 3’ scRNA-seq is grouped by library, however, a submission may contain non-10X data modalities that need to be processed separately.

> An analysis pipeline event handler component will be created that is responsible for partitioning a *data group* into co-processing units, dispatching the pipelines, and tracking completion. This is the responsibility of Analysis.  When analysis results are submitted to ingest, this creates a secondary analysis *data group*.

Reviewer 4: The idea is that these analyses still need to generate their own bundle notifications (as at present) but also need to generate a data group notification? This is definitely going to get us into some intricacies of how analysis will signal those groups, since we can't simply say that everything in a project is in the data group.

Reviewer 2: My understanding is the DSS generates the events when actions are taken on a bundle.  So analysis doesn't directly signal. 

We could put all bundles in a single per project group and let the client sort it out by introspection.  However, this is less obvious and probably a lot of overhead.  We are updated the RFC to make this clearer.

The creation and update of *PROJECT_SUBMISSIONS* *data groups*, both primary and secondary, is the responsibility of the Ingest system.

![Figure 2](../images/0000-multi-data-collection-data-processing-images/flow-of-information.png)
Figure 2: Flow of information in the proposed data model

#### Updates to *data groups*

> Updates to *PROJECT_SUBMISSIONS* must result in updating of the associated *data groups* to trigger reprocessing. When part or the entirety of a project submission is updated the associated data group must be identified and updated with the new bundle versions if bundles have been replaced and/or with additional new bundles if bundles have been added. This is the responsibility of Ingest.  Analysis must handle the update to the *data group* by initiating only the pipelines that have been affected by the update and ensure that the output bundles of this analysis are updates to existing bundles.

Reviewer 4: This means that a PROJECT_SUBMISSION data group must always be the entirety of that project's bundles, as commented above.

Reviewer 2: With project be the unit that is submitted to, then yes, all bundles must be in the PROJECT_SUBMISSION.

Reviewer 11: Currently, submissions are always represented by data from one project. We currently don't have any use cases which would suggest that someone would want to submit data to more than one project at once

Reviewer 4: In fact, this is a design choice we've recently [explicitly made in ingest development](https://docs.google.com/a/ebi.ac.uk/document/d/1qlgIROPK4qESy9-Lwvva1ZWJ3S5EdZ0HWJiGGvt2JRM/edit?disco=AAAADc60OTc) to reduce the high complexity associated with adding new bundles to an existing dataset. Allowing submissions to alter more than one project at once at least doubles the complexity of the problem.

*Data group* updates may result from creating new bundles to add additional types of data to a project, or updating bundles to change metadata or append data to data files (e.g. FASTQ top-up).

#### Implementation feasibility

> Both Analysis and Ingest confirm that implementation of the data_groups bundle proposed here is feasible in principle. Analysis indicates that significant changes need to be made to the existing infrastructure to handle the new type of notification. The generation of secondary data_groups of type PROJECT_SUBMISSION will be generated by analysis submitting to submission and Ingest will identify the bundles included in the secondary bundle.

Reviewer 4: "analysis submitting to submission" -> "analysis making a single submission to Ingest"

Reviewer 8: Sounds like this needs clarification because my understanding was that analysis would do a submission per pipeline, and then ingest would track and identify the bundles to be included in the secondary analysis data group.

Reviewer 4: Right. I'm not sure if it can work the way you describe since ingest would have no way to determine what goes in a group. Analysis would still need to signal this which means a new API mechanism. If you can't submit everything at once then we'd need to look at this in more detail, but that still means you need to track information anyway.

Reviewer 8: Ingest is creating the project submission groups to begin with, so wouldn't ingest know what belongs in the group? Unless you're thinking of the case where we may not actually analyze everything in a group?

Reviewer 2: We just calcified the text as the will be a project group for each level in the data processing pipeline.  However, that doesn't actually invalidate your point, as ingest could look it up.

The underlying issue is that there isn't a 1-to-1 mapping of analysis to primary data bundles, ingest doesn't know the mapping and doesn't know when all of the input bundles have been analyzed.  t somehow has to be told.  In the case of the primary data, the produce of the data tells ingest the bounds of the data.  Analysis will have to do the same.

Sorry ...

Reviewer 8: > The underlying issue is that there isn't a 1-to-1 mapping of analysis to primary data bundles, ingest doesn't know the mapping and doesn't know when all of the input bundles have been analyzed

Ok, that makes sense to me! In the case of a submission containing some data we cannot analyze, is the expected behavior that the secondary data group would only contain the analyzed data and then it would get updated later if the rest of the data can be analyzed?

> Both Analysis and Ingest confirm that implementation of the data_groups bundle proposed here is feasible in principle. Analysis indicates that significant changes need to be made to the existing infrastructure to handle the new type of notification. The generation of secondary data_groups of type PROJECT_SUBMISSION will be generated by analysis submitting to submission and Ingest will identify the bundles included in the secondary bundle.

Reviewer 4: So an analysis submission is not a PROJECT_SUBMISSION data group? In which case, what kind of data group is it?

Also, the question of analysis updates cannot be resolved right now since it depends on whether it makes sense to signal analysis versions using bundle versions. This is pending an Oct 2019 F2F discussion.

Reviewer 2: It is a project submission.  We are going to change the terminology to "project group", since "submission" causes a lot of confusion.

There would be an analysis results projected-scoped data group.  We are updating the text to make this clearer.

The following order of implementation is proposed, accounting for the dependencies between components to ensure uninterupted system functionality:

1) Ingest adds functionality for supporting data_groups and starts using them for singalling submission completion
2) Analysis adds handler for data_groups but continues to use bundle events during a transition period
3) Analysis switches to using data_group events

> 4) Ingests adds support for secondary data_groups

Reviewer 4: I think we can get up to here currently but going beyond requires resolution of the analysis updates RFC.

5) Analyis starts notifying Ingest for the existence of data_groups
6) Downstream components (Matrix Service, Portal) start listening for secondary data_groups

### Unresolved Questions
#### Aspects of the design to be resolved during implementation:
* Where does the *data group* JSON schema belong?  It is not part of the experimental data model, so doesn't fit in the current hierarchy.

* The bundle types for primary vs secondary *data groups* needs to be defined.

#### Aspects of the design to be resolved in the future
* *Data group* concept relates directly to the UX discussion on *experiment data sets*.  This RFC and that work could be unified with a direct mapping between *experiment* and *data groups*.

## Dependencies

> Implementation of this RFC is partly dependent on [RFC: Bundle Types](https://github.com/HumanCellAtlas/dcp-community/pull/86), but approaches that prevent blocking are possible.

Reviewer 4: It is also dependent on the analysis updates RFC

Reviewer 2: I don't think it is blocked by the analysis updates RFC, only that some functionality using this RFC can't be implemented until the analysis updated RFC is implemented.

Reviewer 4: True dat. Partially dependent.

![Figure 1](../images/0000-multi-data-collection-data-processing-images/project-submission-in-project.png)

## Application to analysis pipelines

While this RFC proposes a general mechanism, it was developed in response to the needs of analysis to co-process data that are in multiple assay bundles.  This section uses co-processing of multiple sequencing run from a given library as an example of the use of data groups to identify data that should be processed together.  A key concept behind this proposal is that upstream submission should not be defining how downstream analysis groups inputs for processing.  The analysis pipelines need to be able to group input based on ways that may not be defined at submission time.  New analysis pipelines should be able to group inputs in arbitrary ways without requiring changes to the way data is packaged.

### Identifying data to co-process

> Pipelines that require grouping data from multiple sequencing assays (co-processing) subscribe to *PROJECT_SCOPES* events instead of per-bundle events. Unlike with the assay bundle per pipeline run, this makes the pipeline framework responsible for collecting assay bundles into pipeline runs.  Pipelines that do not require co-processing continue to operate on a per-bundle basis, however they still need to subscribe to the project data group so they can notify ingest when the input *data group* process is complete for use by downstream consumers, such as Azul.

Reviewer 8: I think this would be more accurate if it said `the pipeline service needs to subscribe to the project data group...` 

The difference is that if every _pipeline_ subscribed to "project data group notifications", whenever a project is submitted the same notification would get sent to each pipeline. This creates a lot of duplicated work in terms of inspecting the contents to determine if that pipeline should be run so it would be more efficient to have a single subscription to data group notifications. However, if there's a way to query the project data group bundles without requesting all of the metadata files from the data store then one subscription/pipeline would be fine.

Reviewer 2: thanks, @samanehsan text has been clarified

> ### DCP PR:

Reviewer 12: The main questions that I have upon reading this are high level: 

1. Why does our existing bundle structure not work? I want to be convinced that the bundles for Pe'er and ICA weren't created incorrectly, or created based on faulty metadata or data file headers.
2. Second, why can an "update" in the context of the ongoing AUDR work **not** fix this issue by generating a single bundle containing all files to be processed together?

Reviewer 4: On the second point, at the very least there's currently no way to indicate bundle lineage except by versioning in the datastore. This only allows a single parent for a child.

Reviewer 13: Sorry to re-open this "resolved" conversation. @samanehsan and I have talked about Ambrose's first comment above, i.e. 
>  I want to be convinced that the bundles for Pe'er and ICA weren't created incorrectly, or created based on faulty metadata or data file headers.

Based on my own understanding, we are and will be facing at least 3 problems: 

a. pipelines need to process data span multiple data bundles
b. pipelines need to process data span multiple projects
c. data submitters (e.g. labs) want to submit supplementary data in addition to their original, completely submitted data after a certain time period. 

- The case `c.` is tricky and I believe most of the data platforms will inevitably face a similar issue. I also believe this RFC proposed a good and feasible way to deal with that scenario.

- The case `b.` to me is a perfect candidate to be resolved by this RFC, since we don't have a mechanism to enable pipelines process across projects, unless the submitters making a weird submission with data from multiple projects intentionally.

- When it comes to case `a.`, this RFC pointed out that
    > The current DCP model of processing has a scope of one data bundle, which contain a single assay or analysis. There is no mechanism to know when a group of bundles that need to be processed as a group incomplete and available.

    I'm not super familiar with how the bundles are created from the spreadsheet at the upstream service, but if the pipelines have to process multiple bundles in a single project together, does it mean the algorithm/rule/links resolver that deciding which files(fastqs + metadata) go to which bundle needs to be updated at the first place (prior to DSS), instead of adding extra layers to group the created bundles afterward? It'd be great if someone knows more about the bundle creation mechanism can point out what I missed there if I'm wrong!

Reviewer 2: @rexwangcc thanks for reopening.

I will let @justincc explain the bundling algorithm.  

In my view, bundles are a storage concept, not an experimental concept.   No matter what the bundling policy is, per-bundle processing give ingest control of how analysis experiments are designed.  Analysis should have the ability to construct experiments in any way they see fit without the need to change the storage model.

Reviewer 4: Hey @rexwangcc. Here's [a doc on bundling](https://docs.google.com/document/d/1cGfolHMGEe1IKkXSBS75SAdSri9Np_oW6Z4S9-Ax1QY).

***Leave this blank until the RFC is approved** then the **Author(s)** must create a link between the assigned RFC number and this pull request in the format:*

`[dcp-community/rfc#](https://github.com/HumanCellAtlas/dcp-community/pull/<PR#>)`

# Processing Datasets that Span Multiple Data Collection Runs

## Summary
This RFC proposes a design solution to allow datasets that span multiple data collection runs to be processed in unison as per the scientific requirements. The proposal generalizes to multiple data type modalities, but also provides specific details for 10X V2 3’ single-cell data as an example and to address an immediate need for a data processing solution.

## Author(s)
[Nick Barkas](mailto:barkasn@broadinstitute.org),
[Mark Diekhans](mailto:markd@ucsc.edu)

## Shepherd
-[Nick Barkas](mailto:barkasn@broadinstitute.org)

## Motivation

> Correct data processing requires the initiation of data processing pipelines with all the appropriate input data. The current mechanism of bundle notifications is not effective when data spans multiple bundles because there is no clear set of rules to identify the bundles that need to be co-processed.  Limiting the downstream process to the scope of a single bundle requires the submitter to define the scope of the processing without any foreknowledge of future processes needs.

Reviewer 11: Submitters bear no responsibility for how data is bundled internally. There is a set of rules for how data is bundled that the submitter has no control over.

Reviewer 2: This should be reworded.  Where are the rules documented?  I would assume they still trigger bundling based on the experiment design (per assay??), so there is a connection from what the submitter creates and how ingest creates bundles.  We need to keep this rules-based and not put more burden on the user.

Reviewer 11: @justincc Is there a good place for documentation on how the exporter created bundles right now other than the code?

Reviewer 4: https://docs.google.com/document/d/1cGfolHMGEe1IKkXSBS75SAdSri9Np_oW6Z4S9-Ax1QY/edit. Not complete so updates welcome

Reviewer 4: Ultimately this rules comes from whatever anonymous person made the decision. AFAIK it's not documented anywhere (except in that scratch thing I wrote) though no doubt the crowd could dig up something from 2017 that has got buried under the mulch.

> An immediate need to process 10X V2 scRNA-seq datasets that span multiple bundles exists. The need for co-processing multiple bundles is not however limited to this scenario and extends to any other NGS modality that involves repeated sequencing of the same library. Furthermore the need to co-process data may extend to other data modalities that the project will accept in the future.

Reviewer 11: Can you give a clear example of the immediate need? if this is the library sequencing over multiple lanes/plexes of sequencing there has been an adhoc solution in this instance and the library prep entity provides another solution. In that context, there is no current need for multi-bundle processing. If this is another example, please provide it

Reviewer 2: I am confused.  So is data being bundled by library prep? 

If there is not an immediate needed, that would lower the priority of implementation, how this would be good to have a design in place.  Rebundling primary submissions based on the needs of analysis pipelines is problematic.

Reviewer 11: The ICA and pe'er dataset were fixed in an adhoc manner and the RFC proposed by @malloryfreeberg would enable bundling on library prep rather than cell suspension which would provide resolution 

https://github.com/HumanCellAtlas/dcp-community/pull/87

### Notification Requirements
For a notification to be actionable by analysis, it must fulfill specific requirements. Specifically, the notification must include information about the bundles/data that belong to the same sequencing set and potentially need to be co-processed.

We propose the term **Data Aggregation and Processing Set (DAPS)** to encompass all the different alternative representations of such sets. The only requirement imposed on such these representations is to be reproducible and resolvable to an immutable sets of data. Example representations of DAPS can be:

*   A listing of all relevant bundles along with versions
*   A search query accompanied by a timestamp (to guarantee reproducibility and immutability)
*   Other future representations as project needs evolve (e.g. linking to a particular project bundle version)

> Critically, the submitted DAPS will not explicitly dictate the files that have to be co-processed in pipeline instances as this would require specialized knowledge of the pipelines by the submitter. Instead, **a DAPS indicates that aggregation of data must be performed only from the list of data provided (submission envelope), but the actual aggregation is delegated to the process of analysis initiation, subject to well defined modality-specific metadata requirements**. 

Reviewer 11: How are the rules about how DAPS should be created defined? 

When are those rules invoked to create DAPS?

How does the introduction of DAPS change the data lifecycle in the DCP, a comparative diagram would be really helpful here.

Reviewer 2: This all needs to be fleshed out in the next pass, both in the conceptual model and concrete examples,

Reviewer 11: I will look forward to seeing the improved conceptual and concrete examples

The imposition of specific requirements to metadata is part of this RFC proposal and a key component of the proposed DAPS solution.

This approach provides a generalized signaling mechanism suitable for all foreseeable data types while accounting for data type-specific requirements.

The DAPS approach has common elements with project-level bundles that have been proposed in other contexts, but differs in the following ways:

1. It does not impose any specific aggregation level on the data allowing for more flexibility and accommodating future unforeseen processing scenarios
2. Allows specification of bundles included in a DAPS in many alternative different ways (listing, search, etc…)

> 3. Does not impose the creation of a new bundle type that is tied to a specific data modality

Reviewer 11: Why would a new bundle type be tied to a specific data modality - bundle type definition is based on graph shape not assay type is or there some nuance I am missing?

Reviewer 2: this needs to be reworded in a more abstract manner.

Reviewer 4: More abstract?! I need something more concrete. What is a data modality? ELI5 it for me!

4. Dictates specific metadata requirements that guarantee the correct processing of all foreseeable data modalities
5. Decouples analysis initiation from data representation allowing for future flexibility

## General Metadata Requirements

> In the above framework, analysis pipelines will receive a set of bundles that _potentially_ need to be co-processed. As aforementioned, this signal will not explicitly identify the bundles that need to be co-processed in a single pipeline, instead, it will serve as a notification **indicating that any co-processing is to be done only within the boundaries of this set of files (submission envelope). **

Reviewer 3: What do you mean by "submission envelope" here? That term has a pretty specific meaning the DCP.

Reviewer 2: will deleted.

Reviewer 4: Yeah, I wouldn't tie this into the idea of submission envelopes. A submission can be parts of a dataset that the author would consider must be processed together.

> Analysis can initiate multiple pipeline invocations as a result of a single DAPS being submitted.

Reviewer 11: How and why would you need to /want to?

Reviewer 2: The goal is to not have one to one restriction on analysis to primary data.  Multiple algorithms or types of analysis are allowed.

why would we want to restrict the systems in this way?

Reviewer 11: I understand not wanting to unduly restrict the system at the same time I also don't want to overcomplicate the system to enable a rare case which can be solved using more simple means

Reviewer 4: I don't see any complexity here. If one pipeline can be invoked then it's trivial to invoke multiple. Maybe not even worth saying.

The actual grouping of the files is to be performed by analysis via a search in the metadata of the bundles submitted following well-defined modality-specific set of rules. For this reason, the metadata must satisfy the following metadata requirements (MR):

Furthermore, if part of the data submitted in a DAPS has already been processed the Analysis infrastructure will only re-process the parts that have not been modified.

> 1. The data modality of each data file included in the DAPS must be clearly identifiable (MR1)

Reviewer 11: What do you mean by data modality?

Reviewer 2: data modality == sequencing protocol.

Reviewer 11: Does that means the DAPS will need metadata above and beyond what is recorded in the DCP scientific metadata files?

> 2. Data that need to be processed together must be clearly identifiable by a modality-specific set of metadata variables that will group the files according to processing requirements. (MR2)

Reviewer 11: Please give an example of specific metadata fields which could be used to trigger an analysis?

Reviewer 2: examples are in a later section of the document

Reviewer 11: It would be great to see them here, having to flip back and forth in a document to understand its meaning makes it very difficult to process.

> 3. Among the data that needs to be processed together, the metadata must provide sufficient information to correctly order, pair or otherwise establish correspondence and type of the input files. (MR3)

Reviewer 11: Please give a specific example of what is meant by ordering of files

## Proposed Signalling Implementation

> The DAPS proposal above only sets specific requirements for the signal to initiate data analysis and minimum requirements for the associated metadata. We hereby provide a specific recommendation for implementation via bundle notifications.

Reviewer 11: What is the purpose of inventing a new type of aggregate structure in the DSS. Why not create another bundle type? - this seems to be adding complexity for no clear advantage

Reviewer 2: DAPS is a bundle type that is used to group other bundles.  If this isn't clear, we need to update the text and for sure add an image.

Reviewer 11: That isn't clear at all. Some diagrams would be a great help here

We recommend that the data store (DSS) is used to store DAPS as a new bundle type of **“DAPS Manifest”**, and that the current search-based notification system is adjusted so that instead of listening for new primary bundles, analysis listens for new DAPS bundles that are exclusively used for the initiation of pipelines. This implementation has two major advantages: (1) the DAPS are permanently saved at the main storage site for the HCA project (DSS) so the analyses can be reproducibly triggered and (2) The existing notification mechanisms are reused, avoiding the need for implementation of a new out-of-band notification system.

We propose that as an initial implementation approach DAPS are used at set data aggregation levels (e.g. sample, project) but critically the implementation must be flexible to arbitrary data aggregation levels to accommodate future project needs. We propose that triggering is initially done manually by data wranglers and/or ingest, but we envisage transition to automatic triggering in the future.

> ## Example: Analysis Implementation for a DAPS

Reviewer 3: I think it would be helpful if this example had more detail. Most of this RFC is pretty abstract, deliberately eliding implementation details. But, it's hard to understand what exactly you have in mind without a worked example. For instance, in this 10x case who would create the DMB and how would they do that? What if additional sequencing was done later? Where does DM parsing code live, and who writes/maintains it? What's the lifecycle of a DMB?

The following pseudocode presents a proposed implementation of the processing of an incoming DAPS by analysis:
```

> Partition the datasets by data modality (MR1)

Reviewer 11: What do the MRN acronyms mean here?

Reviewer 2: MRn is an identifier  described in General Metadata Requirements above

Reviewer 11: This acronym use doesn't' seem helpful but instead confusing, they are only used in two places in the document very close to each other, it isn't clear how they help understanding

For each data modality:
    Use the modality-specific criteria to identify pipeline invocations (MR2)
    For each invocation:
        order the data as required (MR3) and launch pipeline invocation
```

> Note that the above approach does not require all the data in a submitted DAPS to be of the same modality to be processed correctly. For example a DAPS at the project level can contain imaging and 10X datasets and these will be correctly processed in an independent manner. 

Reviewer 11: If being present together in the same project does not guarantee the need for uniform and grouped processing why create a project-level DAPS? what function does it serve?

Reviewer 2: DAPS is a signaling mechanism that a group of bundles is available.  Right now it is impossible to know when data is complete beyond a bundle.  There is no way for ingest, or any other user, to know if a project or some subset is complete.

Reviewer 11: I suspect with some lifecycle diagrams and pointers to how and when the DAPS bundle would be created it will be easier to see how this is different from a bundle which contains everything from a project

Furthermore if part of the data submitted in a DAPS has already been processed the Analysis infrastructure will only re-process the parts that have not been modified.

## Metadata Requirements for 10X V2 3’ scRNA-seq datasets
The specific implementation of the above metadata requirements for 10X V2 3’ RNA-seq datasets is presented here as an initial use case and an example. 10X V2 3’ data are the most common type of data currently deposited in the DSS and therefore represent a widely applicable use case.

Requirement (1) above can be satisfied by having a globally unique ‘Library ID’ variable in the metadata that identifies the Illumina sequencing library from which each input file originates from. More details on the requirements and a specific implementation of this solution can be found currently under consideration here: [PR](https://github.com/HumanCellAtlas/dcp-community/pull/87), [document](https://github.com/HumanCellAtlas/dcp-community/blob/mfreeberg-rfc-library-preparation/rfcs/text/0000-rfc-library-preparation.md).

> Requirement (2) can be satisfied in two alternative ways: (a) sufficient metadata can be provided as outlined in table 1 below to identify all file triplets, or (b) a globally unique triplet field can be used to identify file correspondence by the submitter. We propose the latter approach as collecting all the required metadata may not be feasible in all cases and places an undue burden on the submitter in the common scenarios where only one triplet is submitted.

Reviewer 11: Globally unique file set id - how do you propose that is created? what information is needed to understand what a fileset is? how often is that info needed? can it be inferred and only requested if the submitter has a complex library sequencing strategy?

We propose that the metadata required to automatically identify the triplets (Table 1) is made optional and users are encouraged to submit these fields as it can help identify errors and also can in some cases provide important covariates for scientific analyses.

**Table 1: Fields required for identification of fastq file correspondence**
<table>
  <tr>
   <td><strong>Required Fields</strong>
   </td>
   <td><strong>Description</strong>
   </td>
   <td><strong>Justification</strong>
   </td>
  </tr>
  <tr>
   <td>Run Date and Time
   </td>
   <td>ISO Date and time at which the run was performed
   </td>

>    <td>This is for the rare scenario where flowcells are re-used and ensures that each run is processed as if it were a separate flowcell

Reviewer 11: Can you give a real example of this that triggers this need or is it only a hypothetical? are there data in the same submission which would have used the same lane of the same flowcell meaning this DateTime would be needed? That seems sufficiently unlikely to make this a big ask for submitters

   </td>
  </tr>
  <tr>
   <td>Flowcell ID
   </td>
   <td>The Illumina identifier of the flowcell id
   </td>
   <td rowspan="3" >The unique combination of Flowcell ID, Lane and Barcode uniquely identifies any multiplexed library on an Illumina Flowcell
   </td>
  </tr>
  <tr>
   <td>Lane in Flowcell
   </td>
   <td>The lane in the flowcell
   </td>
  </tr>
  <tr>

>    <td>Barcode of Library

Reviewer 11: Why is the library barcode needed? we only take demuxed fastq files? how does that information help?

   </td>
   <td>Barcode (or concatenation of the barcodes) of the library
   </td>
  </tr>
</table>

## Metadata Requirements for SS2 scRNA-seq datasets and provision for plate-based analyses
The above metadata requirements are also compatible with SS2 analyses and can support plate based-analyses. Cells that have been sequenced more than once (a rare, but plausible scenario for SS2) can be identified by using a library identifier as presented in this RFC, currently under review.

> Furthermore, this approach can be extended to support plate-based processing. By ensuring that a plate field is provided for each cell, multiple cell bundles can be aggregated into a single plate-based workflow. A DAPS can contain one or more plates. Provision must be made for any cells not associated with a plate identifier, or otherwise, the plate identifier must be a compulsory field.

Reviewer 11: Plate name is already required for SS2

## Pros and Cons of DAPS Approach

Pros
*   Avoids potentially very expensive run-away queries
*   Ensures that analyses are triggered once with the appropriate datasets
*   Is flexible and can accommodate all foreseeable data modalities
*   Can immediately resolve processing challenges with 10X datasets
*   Can provide a mechanism for moving to plate-based processing for SS2 in a seamless manner
*   Does not preclude future automation of pipeline initiation when a mature system state is achieved
*   Re-uses existing notification infrastructure
*   Is permanently saved in the DSS and analyses are reproducible
*   Provides a natural place to provide analysis-specific parameters

Cons
*   Analysis need to modify pipeline initiation scripts

> *   Initial implementation has human time cost of triggering the pipeline runs

Reviewer 11: Why would this introduce human triggering of pipeline runs? when is a human needed in this process?

Reviewer 2: this should be made a requirement on ingest to generate the signal.

Reviewer 4: At least initially, Ingest could generate a signal that a whole project is ready for analysis when a submission completes. I think it's an open question how you'd signal any other kind of grouping like a dataset.

> *   Requires the human who is triggering the pipeline to know what criteria need to be satisfied to trigger (i.e. the project is complete)

Reviewer 11: Requiring a human to "know" when something is complete in order for this process to work feels like a very big risk for two reasons, the first is depending on who this human is, they might not have enough info to know, the second is it requires a human to proactively run the process which feels like it should be avoidable if there are clear rules on how the DAPS get created

Reviewer 2: that is why it is not desirable.  the current use case would be the completion of a submission, which we now realize ingest already knows.   We will discuss other cases that may need more human specification in the next revision.

Reviewer 4: A problem here is that a submission could be extremely small and only make up part of the grouping that should be analyzed together. One could make a submission just to add a single FASTQ, for example.

## Alternative Approaches 

The following section summarizes possible alternative approaches for ensuring correct processing, along with the pros and cons of each. These approaches are not recommended by this RFC but are listed for comparison purposes.

*   Placing all the data that needs to be co-processed in a single bundle and continue to use bundle notifications
    *   Pros
        *   Essentially no extra work from infrastructure required
    *   Cons

>         *   Wranglers are required to package data in such a way to satisfy the requirements of the analysis, the details of which they should not be aware

Reviewer 3: Don't they need to know these details to create a manifest bundle?

Reviewer 11: Wranglers don't create the manifest bundle, the ingest service does that. That should also be true here, manual creation steps in this process seems like a bad plan

Reviewer 2: that should say "ingest" not "wrangers"

Reviewer 4: Beyond an aggregation like project somebody still needs to specify the groupings. either manually or as rules for ingest to execute to generate DAPs bundles.

>         *   Wranglers are required to package data in such a way to satisfy the requirements of the analysis, the details of which they should not be aware

Reviewer 11: Why the desire to avoid wranglers understanding what makes reasonable input for an analysis process? Surely they need to understand this to ensure they are having the correct conversations with submitters to get the right information in the first place? It isn't clear why that understanding is a bad thing, certainly at this stage in the project

Reviewer 2: There is no desire for wrangler avoiding understanding anything.  there is a requirement that wranglers do not have to package the data based on analysis pipelines that may not even exist when the data is ingested.  This needs to be made clearer
